### ISSUE TITLE
cloudinit with cicustom is deleted

### ISSUE THREAD

**Issue by electropolis** [Telmate/terraform-provider-proxmox#956]
My problem is a structured case that seemed to exist in parallel with the problems of people who do not use cicustom. It was created to help to summarise all efforts of trying to resolve it [here](https://github.com/Telmate/terraform-provider-proxmox/issues/901#issuecomment-1952846244). So people ask me to create own issue to address my case with `cicustom` and `cloudinit`.
Is mostly related with removing CloudInit drive, and creating ide2, which is a plain CD-ROM, during the provisioning process. For people who DON'T use `cicustom` instead do use `causer,` `cipassword` have a workaround to:
1) Avoid creating template with Cloudinit as ide2
or
2) Create template with Cloudinit as ide3
or
3) Set boot order to `scsi0;ide2` or `scsi0;ide3` in terraform script. 

None of above works in my case. 

The problem which I believe is happening is related with changing the VM configuration during the provisioning process.
On sample video Here I made changes in terraform that fails when setting the boot order
` boot        = "order=scsi0;ide3"`

That crashes with error
```
╷
│ Error: error updating VM: 500 invalid bootorder: device 'ide3' does not exist', error status: {"data":null} (params: map[agent:1 bios:seabios boot:order=scsi0;ide3 cicustom:user=local:snippets/user-data_vm-srv-app-1.yaml,network=local:snippets/network-config_vm-srv-app-1.yaml cores:4 cpu:host delete:ide3 hotplug:network,disk,usb ide2:none,media=cdrom kvm:true memory:8192 name:srv-app-1 net0:virtio=00:1E:67:01:10:01,bridge=skynet numa:false onboot:false scsi0:vms:vm-100-disk-0,replicate=0,ssd=1 scsihw:virtio-scsi-pci sockets:1 tablet:true vmid:100])
│
│   with proxmox_vm_qemu.cloudinit["srv-app-1"],
│   on main.tf line 1, in resource "proxmox_vm_qemu" "cloudinit":
│    1: resource "proxmox_vm_qemu" "cloudinit" {
│
╵
```
And in WebUI looks like this

https://github.com/Telmate/terraform-provider-proxmox/assets/16556636/d0305348-3065-402d-ada2-3b006754ff99

On the video you see that when VM updates with bigger amount of memory, cores, disk capacity there is additional CD-ROM that appears. By the way why it's forcing to add that device (something like that wasn't in PVE 7) and suddenly cloudinit values get crossed which means it gets deleted. So basically doesn't matter if you avoid cloudinit when creating template and then you create vm from that template it still avoids cloudinit and setting different boot order as you see here on picture.

<img width="655" alt="image" src="https://github.com/Telmate/terraform-provider-proxmox/assets/16556636/dec01f52-55be-436a-8574-b14ee75941d6">

Instead of `ide3` it ignores that from terraform `  boot        = "order=scsi0;ide3"` and sets `ide2` instead. 
Everytime `ide2` is forced to be set because it says that `ide3` doesn't exists (from above error). 

For me it ruins whole concept of using pve8 when script doesn't seems to work at all. Can you do in code when setting `cloudinit_cdrom_storage = "local-lvm"` to avoid using `ide2` as that damn `CDROM` ?
I hope this the is best summarise I can provide. I appriciate any help here.

My code is here:
Terraform: https://github.com/sonic-networks/terraform/tree/master/proxmox
Ansible for generating the templates: https://github.com/sonic-networks/ansible/tree/master/proxmox

**Comment by hestiahacker** [comment]
I attempted to deploy a VM using the terraform files that you linked to and I'm getting errors:

```

╷
│ Error: No value for required variable
│ 
│   on provider.tf line 10:
│   10: variable "TF_PROXMOX_API_ID" {}
│ 
│ The root module input variable "TF_PROXMOX_API_ID" is not set, and has no default value. Use a -var or -var-file command line argument to provide a value for this variable.
╵
╷
│ Error: No value for required variable
│ 
│   on provider.tf line 11:
│   11: variable "TF_PROXMOX_API_SECRET" {}
│ 
│ The root module input variable "TF_PROXMOX_API_SECRET" is not set, and has no default value. Use a -var or -var-file command line argument to provide a value for this variable.
╵
╷
│ Error: No value for required variable
│ 
│   on variables.tf line 32:
│   32: variable "hosts" {
│ 
│ The root module input variable "hosts" is not set, and has no default value. Use a -var or -var-file command line argument to provide a value for this variable.
╵
```

After commenting out most of what's in provider.tf, I get errors about the host not being defined...

```
╷
│ Error: No value for required variable
│ 
│   on variables.tf line 32:
│   32: variable "hosts" {
│ 
│ The root module input variable "hosts" is not set, and has no default value. Use a -var or -var-file command line argument to provide a value for this variable.
╵
```

Can you post an simple example that will reproduce this issue? If not that, can you give us a command that specifies all of these variables on the command line (e.g. with `-var` as the error messages suggest) so people can reproduce this issue outside of your environment?

**Comment by electropolis** [comment]
@hestiahacker  the first issues are obvious. No auth added to env. So in look for `terraform_pve1` that has those variables added and change the values of them as it's adjusted to work with macOS keychain. When it comes to `hosts` . Sorry I didn't pushed the `tfvars`  file, but you have in README information what is the structure of it or even in `variables.tf` there is declaration of that variable

```
variable "hosts" {
  type = map(object({
    hostname = string
    cpu      = number
    ram      = number
    hdd      = string
    template = string
    macaddr  = string
    ip       = string
    bridge   = string
  }))
}
```
and example of it you can see in [README](https://github.com/sonic-networks/terraform/tree/master/proxmox)

**Comment by hestiahacker** [comment]
I'm sure I could spend more time making more changes to your example to get it to work outside of your environment, but I'm asking that you meet me part way. I'm already volunteering my time and labor to do testing, diagnosing issues, coordinating bug fixes, updating the documentation and so forth.

The only thing I ask is that people reporting issues give me a minimal, portable example to demonstrate the issue. If you are not willing to do that, then I'm not going to prioritize this ticket. I'll still work on it, but I'll get to it whenever I get to it.

Lets work together to get this fixed for you.

**Comment by electropolis** [comment]
> I'm sure I could spend more time making more changes to your example to get it to work outside of your environment, but I'm asking that you meet me part way. I'm already volunteering my time and labor to do testing, diagnosing issues, coordinating bug fixes, updating the documentation and so forth.
> 
> The only thing I ask is that people reporting issues give me a minimal, portable example to demonstrate the issue. If you are not willing to do that, then I'm not going to prioritize this ticket. I'll still work on it, but I'll get to it whenever I get to it.
> 
> Lets work together to get this fixed for you.

You know what? I don't understand what do you want from me know. You pasted some CLI outputs. So I explained you that these are very simple issues to understand and you didn't authorised yourself against your proxmox. I don't know how you do it, but I gave you a clue that `terraform_pve1` has required `TF_VAR_` variables that need to be loaded into OS `env`. That was the first case. Second again, I gave you information that in the README file there is explanation how to create the map object which that is obviously declared in `variables.tf` . I mean , this is just a simple knowledge that anyone should have when working with terraform. If you don't understand it, or you have a lack a of some basic knowledge that's not a problem just say it. It's not a shame. But I obviously assume that you are well skilled in terraform so I answered very shortly what didn't you do. So what do you expect me to do ? Explain you the basic steps how to get it run? I already did. And I think that's the standard way of working with terraform. 

>The only thing I ask is that people reporting issues give me a minimal, portable example to demonstrate the issue. If you are not willing to do that, then I'm not going to prioritize this ticket. I'll still work on it, but I'll get to it whenever I get to it.

Im sorry but I don't understand that sentence. What is minimal, portable example?  I gave you the code. I gave you the screenshots, You have it all. So the only thing what you need to do know is to clone that code, auth yourself against your proxmox server and defined that `hosts` which you have a basic example in the README file that I gave you. and run it. If there are any other issues, problems address them directly , "where to put that hosts map?" and so on. Because I don't know what you don't understand. You need to define exactly what you want.

**Comment by hestiahacker** [comment]
What I mean by a simple, portable example is one that has few parts or features and is not complicated or elaborate. Something like this: https://github.com/Telmate/terraform-provider-proxmox/issues/704#issuecomment-1925585063

Or the template I created for reporting bugs here: https://github.com/Telmate/terraform-provider-proxmox/pull/950/files

These are much smaller than what you provided, and only require a few lines to adapt to anyone's environment. The template I created has just 7 lines that might need changed and they're all in one place. There's no hardcoded links to internal IP addresses of your PVE server that we have to rip out, no unnecessary host maps, terraform_pve1 files that need changed, readme files to consult, and so on.

If you'd like to provide such an example, it'd save time for everyone who is willing to help you with this issue. It would also avoid a bunch of comments back and forth on the ticket about the definition of the word "simple", trying to debate whether people volunteering to help you should be putting in more effort, or accusing them of having communication problems.

**Comment by electropolis** [comment]
> These are much smaller than what you provided
But what is smaller? I described the problem. Like anyone here did. No one is using your template for anything. Second thing is that you took the effort and you cloned the repo and wanted to use it and encountered very basic dilemma. How to run the code which is quite obvious. There is no rocket science here. And I spend my time to explain you simple as it's already described. I mean if someone has difficulties how to run the code - look into README ? As simple as that. But you started to showing some problems and pushing me to provide you with additional details and use YOUR template. 

> There's no hardcoded links to internal IP addresses of your PVE server that we have to rip out, no unnecessary host maps, terraform_pve1 files that need changed, readme files to consult, and so on.

So how you wanna test this configuration  if you wanna put some modification. Thats not serious. You wanna me to adapt the code to your needs. Sorry but not gonna happen as this is not solving issues but attempting to override them with your solution and letting me know "here is the working solution, just change to mine". 
If you don't wanna help so don't do it. My code looks like this. If you have difficulties with understanding it and replace some values with yours (it's easy) 
because there is no hardcoded IP , where? 
Here -> ` ip       = "10.xx.xx.xx"` ? 
But this the same where you provide it with `ipconfig0 = "ip=${var.ip_address}/${var.cidr},gw=${var.gateway}"`  , the logic is the same.  You can even put there whatever you want, doesn't matter really because network here is irrelevant as the problem is with cloudinit device. 

And of course if you wanna do something, test someone code you need to be ready that you gonna have to adjust your parameters to that code not the other way around. 

And I can't provide any of that example because you already have the information what to put where. So add there your proxmox bridge name, IP (whatever you like) na `terraform_pve1` your auth of doesn't matter export it from CLI directly , I don't care until the required `TF_VAR` are added which are the one declared in `provider.tf`
```
variable "TF_PROXMOX_API_ID" {}
variable "TF_PROXMOX_API_SECRET" {}
```

I have this strange feeling that you don't understand terraform and trying to force me do adjust something for you level of knowledge by adding a template. But how can I add a template when if you wanna test my config you HAVE to work on my template not yours. So like I said, and don't get me wrong, if you don't understand something ASK. If not, so you can't help me at all. Someone else who understand how to authenticate itself by loading variables into OS env - which is a standard common way of testing something, I think. Maybe you are doing it differently. I don't have a clue. And also knows how to do copy/paste from README into file `<name>.auto.tfvars` and changing variables for it needs. I said _it needs_ because I don't know which config you have (the name of the bridge, name of the template and so on). All those things you need to set it by your own, I can't help you with that because it's impossible. But for just looking at the code someone who has some knowledge should understand how to deal with that code as it's very common. If not, than I'm willing to help but by explaining, helping to configure and not changing my code to some kind of template. 

Im not gonna debate on what is simple or not. And not gonna force someone to help me. You wanted to help me, you said to raise a ticket ? I did. Explained everything - and you still don't understand and ask me to refactor my config so that you will understand it. I mean that doesn't make sense at all because then you will not test the case here. And I'm not accusing anyone of having communication problems. I'm referring to our communication problems. Right now this issue generates another issue between your understanding the code and effort to work with that code and debating with me that doesn't helps and it's waste of time. Finally - you wanna help me? Ask me what you don't understand as you already took effort and cloned the repo but had some simple issues (I mean having problem with empty variables is rather a simple issue or isn't ?) which I already explained you. And still you are debating and forcing me to do something. Sorry I won't. Thank you for your effort.

**Comment by Tinyblargon** [comment]
This should be resolved in the latest build #959 has an example.

**Comment by Tinyblargon** [comment]
Fixed by v3.0.1-rc2
